### PR TITLE
Fixes scheduled execution errors on rollback and some refactoring

### PR DESCRIPTION
**Pull request by AdoAdoAdo**

### CODE REVIEW

File: process/block/baseProcess_test.go [hunk 1]
```diff
@@ -365,7 +365,8 @@ func CreateMockArguments(
 			Version:                      "softwareVersion",
 			HistoryRepository:            &dblookupext.HistoryRepositoryStub{},
 			EpochNotifier:                &epochNotifier.EpochNotifierStub{},
-			ScheduledTxsExecutionHandler: &testscommon.ScheduledTxsExecutionStub{},
+			ScheduledTxsExecutionHandler: &testscommon.ScheduledTxsExecutionStub{
+			},
```
Comment (SebastianMarian) [inline process/block/baseProcess_test.go:369]: Could be on the same line

File: process/block/preprocess/scheduledTxsExecution.go [hunk 2]
```diff
@@ -196,6 +211,12 @@ func getAllIntermediateTxsAfterScheduledExecution(
 
 // GetScheduledSCRs gets the resulted SCRs after the execution of scheduled transactions
 func (ste *scheduledTxsExecution) GetScheduledSCRs() map[block.Type][]data.TransactionHandler {
+	mapScheduledSCRs, _ := ste.createCopyScheduledSCRs()
+
+	return mapScheduledSCRs
+}
+
+func (ste *scheduledTxsExecution) createCopyScheduledSCRs() (map[block.Type][]data.TransactionHandler, int) {
```
Comment (SebastianMarian) [inline process/block/preprocess/scheduledTxsExecution.go:219]: This method is used only in GetScheduledSCRs method above which has only two lines of code. I think that this code could be merged there and also for the debugging scope we could keep the logs as they were before.

File: process/block/preprocess/scheduledTxsExecution.go [hunk 3]
```diff
@@ -285,6 +305,96 @@ func (ste *scheduledTxsExecution) HaveScheduledTxs() bool {
...
+	rootHash, _, err := ste.getScheduledRootHashAndSCRsForHeader(headerHash)
+	return rootHash, err
+}
+
+// RollBackToBlock rolls back the scheduled txs execution handler to the given header
+func (ste *scheduledTxsExecution) RollBackToBlock(headerHash []byte) error {
+	scheduledRootHash, mapScheduledSCRs, err := ste.getScheduledRootHashAndSCRsForHeader(headerHash)
+	ste.SetScheduledRootHashAndSCRs(scheduledRootHash, mapScheduledSCRs)
+
+	return err
```
Comment (SebastianMarian) [inline process/block/preprocess/scheduledTxsExecution.go:321]: return nil here and if err != nil { return err } before L319 ?

File: process/block/preprocess/scheduledTxsExecution.go [hunk 4]
```diff
@@ -285,6 +305,96 @@ func (ste *scheduledTxsExecution) HaveScheduledTxs() bool {
...
+}
+
+func (ste *scheduledTxsExecution) SaveState(headerHash []byte) {
+	scheduledRootHash := ste.GetScheduledRootHash()
+	mapScheduledSCRs := ste.GetScheduledSCRs()
+	if ste.HaveScheduledTxs() {
```
Comment (SebastianMarian) [inline process/block/preprocess/scheduledTxsExecution.go:327]: Maybe this method could could be removed from interface and replaced here with the check done inside it

File: process/block/preprocess/scheduledTxsExecution.go [hunk 5]
```diff
@@ -285,6 +305,96 @@ func (ste *scheduledTxsExecution) HaveScheduledTxs() bool {
...
+		return nil, nil, err
+	}
+
+	scheduledRootHash := scheduledSCRs.RootHash
+	txHandlersMap := scheduledSCRs.GetTransactionHandlersMap()
+
+	return scheduledRootHash, txHandlersMap, nil
+}
+
+func (ste *scheduledTxsExecution) getMarshalizedScheduledRootHashAndSCRs(
```
Comment (SebastianMarian) [inline process/block/preprocess/scheduledTxsExecution.go:382]: Is this used anywhere? Maybe replace L328-L342 with the call of this method?

File: process/interface.go [hunk 6]
```diff
@@ -1113,7 +1113,10 @@ type ScheduledTxsExecutionHandler interface {
 	Execute(txHash []byte) error
 	ExecuteAll(haveTime func() time.Duration) error
 	GetScheduledSCRs() map[block.Type][]data.TransactionHandler
-	SetScheduledSCRs(mapSCRs map[block.Type][]data.TransactionHandler)
+	SetScheduledRootHashAndSCRs(rootHash []byte, mapSCRs map[block.Type][]data.TransactionHandler)
+	GetScheduledRootHashForHeader(headerHash []byte) ([]byte, error)
+	RollBackToBlock(headerHash []byte) error
+	SaveState(headerHash []byte)
 	GetScheduledRootHash() []byte
```
Comment (SebastianMarian) [inline process/interface.go:1120]: Are GetScheduledRootHash and SetScheduledRootHash still used outside the component? If not we can remove them from interface and made them private there.

File: process/sync/baseSync.go [hunk 7]
```diff
@@ -740,13 +740,9 @@ func (boot *baseBootstrap) rollBack(revertUsingForkNonce bool) error {
...
 
-		scheduledRootHash, mapScheduledSCRs, err := process.GetScheduledRootHashAndSCRs(prevHeaderHash, boot.store, boot.marshalizer)
+		err = boot.scheduledTxsExecutionHandler.RollBackToBlock(prevHeaderHash)
 		if err != nil {
-			boot.scheduledTxsExecutionHandler.SetScheduledRootHash(prevHeader.GetRootHash())
-			boot.scheduledTxsExecutionHandler.SetScheduledSCRs(make(map[block.Type][]data.TransactionHandler))
-		} else {
-			boot.scheduledTxsExecutionHandler.SetScheduledRootHash(scheduledRootHash)
-			boot.scheduledTxsExecutionHandler.SetScheduledSCRs(mapScheduledSCRs)
+			boot.scheduledTxsExecutionHandler.SetScheduledRootHashAndSCRs(prevHeader.GetRootHash(),make(map[block.Type][]data.TransactionHandler))
```
Comment (SebastianMarian) [inline process/sync/baseSync.go:745]: Add a space after ,

File: process/block/baseProcess_test.go (hunk 1 above)
Comment (AdoAdoAdo) [reply process/block/baseProcess_test.go:369]: done

File: process/block/preprocess/scheduledTxsExecution.go (hunk 2 above)
Comment (AdoAdoAdo) [reply process/block/preprocess/scheduledTxsExecution.go:219]: yes, previously I was using it in two places, removed, and added back the logs

File: process/block/preprocess/scheduledTxsExecution.go (hunk 3 above)
Comment (AdoAdoAdo) [reply process/block/preprocess/scheduledTxsExecution.go:321]: done

File: process/block/preprocess/scheduledTxsExecution.go (hunk 4 above)
Comment (AdoAdoAdo) [reply process/block/preprocess/scheduledTxsExecution.go:327]: done

File: process/block/preprocess/scheduledTxsExecution.go (hunk 5 above)
Comment (AdoAdoAdo) [reply process/block/preprocess/scheduledTxsExecution.go:382]: yes forgot to use it in SaveState

File: process/interface.go (hunk 6 above)
Comment (AdoAdoAdo) [reply process/interface.go:1120]: yes, still used from baseProcess and baseSync

File: process/sync/baseSync.go (hunk 7 above)
Comment (AdoAdoAdo) [reply process/sync/baseSync.go]: done

**Review by SebastianMarian — APPROVED** [review]

**Review by raduchis — APPROVED** [review]

**Pull request merged**